### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/EllipticCurve/Jacobian): add API for equations and nonsingularity

### PR DESCRIPTION
**Pull request by Multramate**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean [hunk 1]
```diff
@@ -86,240 +86,329 @@ local macro "pderiv_simp" : tactic =>
     pderiv_X_of_ne (by decide : (2 : Fin 3) ≠ 0), pderiv_X_of_ne (by decide : (0 : Fin 3) ≠ 2),
     pderiv_X_of_ne (by decide : (2 : Fin 3) ≠ 1), pderiv_X_of_ne (by decide : (1 : Fin 3) ≠ 2)])
 
-variable {R : Type u} [CommRing R] (W : Jacobian R)
+variable {R : Type u} [CommRing R] {V : Jacobian R} {F : Type u} [Field F] {W : Jacobian F}
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:89]: Why not `F : Type v`? Are there definitely no results which use both R and F here?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean [hunk 2]
```diff
@@ -86,240 +86,329 @@ local macro "pderiv_simp" : tactic =>
...
+    W.a₁ * P x * P y / P z ^ 5 * div_self hPz + W.a₃ * P y / P z ^ 3 * div_self (pow_ne_zero 3 hPz)
+      - W.a₂ * P x ^ 2 / P z ^ 4 * div_self (pow_ne_zero 2 hPz)
+      - W.a₄ * P x / P z ^ 2 * div_self (pow_ne_zero 4 hPz) - W.a₆ * div_self (pow_ne_zero 6 hPz)
+
+variable (V) in
 /-- The proposition that a point representative $(x, y, z)$ lies in `W`.
 In other words, $W(x, y, z) = 0$. -/
 def Equation (P : Fin 3 → R) : Prop :=
-  eval P W.polynomial = 0
+  eval P V.polynomial = 0
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:206]: Do you think `IsOnCurve` would be a better name for this definition than `Equation`? Maybe that's not in scope for this PR...

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean [hunk 3]
```diff
@@ -86,240 +86,329 @@ local macro "pderiv_simp" : tactic =>
...
 
-lemma equation_smul_iff (P : Fin 3 → R) (u : Rˣ) : W.Equation (u • P) ↔ W.Equation P :=
-  have (u : Rˣ) {P : Fin 3 → R} (h : W.Equation P) : W.Equation <| u • P := by
-    rw [equation_iff] at h ⊢
-    linear_combination (norm := (simp only [smul_fin3_ext]; ring1)) (u : R) ^ 6 * h
-  ⟨fun h => by convert this u⁻¹ h; rw [inv_smul_smul], this u⟩
+lemma equation_some (X Y : R) : V.Equation ![X, Y, 1] ↔ V.toAffine.Equation X Y := by
+  simp only [equation_iff, Affine.equation_iff', fin3_def_ext]
+  congr! 1
+  simp only [one_pow, mul_one]
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:233]:
```suggestion
  simp only [equation_iff, Affine.equation_iff', fin3_def_ext, one_pow, mul_one]
```
`congr!` doesn't seem to be needed.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean [hunk 4]
```diff
@@ -86,240 +86,329 @@ local macro "pderiv_simp" : tactic =>
...
-  simp only [W.toAffine.nonsingular_iff, equation_some, and_congr_right_iff,
-    W.toAffine.equation_iff, ← not_and_or, not_iff_not, one_pow, mul_one, Iff.comm, iff_self_and]
-  intro h hX hY
-  linear_combination (norm := ring1) 6 * h - 2 * X * hX - 3 * Y * hY
-
-lemma nonsingular_smul_iff (P : Fin 3 → R) (u : Rˣ) : W.Nonsingular (u • P) ↔ W.Nonsingular P :=
-  have (u : Rˣ) {P : Fin 3 → R} (h : W.Nonsingular <| u • P) : W.Nonsingular P := by
-    rcases (W.nonsingular_iff _).mp h with ⟨h, h'⟩
-    refine (W.nonsingular_iff P).mpr ⟨(W.equation_smul_iff P u).mp h, ?_⟩
+  V.Equation P ∧ (eval P V.polynomialX ≠ 0 ∨ eval P V.polynomialY ≠ 0 ∨ eval P V.polynomialZ ≠ 0)
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:311]: I think `IsNonsingular` is a better name for this but again this might not be for this PR. It's probably also the wrong definition for a general ring: one should instead probably ask that the evaluations are units rather than nonzero. Is this an issue?

**Review by kbuzzard — APPROVED** [review]
This looks great to me! I've left some comments on both this and #9433 .

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean (hunk 1 above)
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:89]: Oops, thanks!

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean (hunk 2 above)
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:206]: Sure, I plan to rehaul both `Equation` and `Nonsingular` in a separate future PR. Their names are not ideal (e.g. I think `Nonsingular` should be `IsNonsingular` according to style conventions), and their definitions can be generalised to be correct over more rings (e.g. PIDs).

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean (hunk 3 above)
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:233]: Thanks!

**Comment by kbuzzard** [comment]
Thanks!

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean [hunk 5]
```diff
@@ -86,240 +86,328 @@ local macro "pderiv_simp" : tactic =>
...
+lemma nonsingular_zero [Nontrivial R] : V.Nonsingular ![1, 1, 0] := by
+  simp only [nonsingular_of_Z_eq_zero, equation_zero, true_and, fin3_def_ext, ← not_and_or]
+  exact fun h => one_ne_zero <| by linear_combination (norm := ring1) h.1 - h.2.1
 
-lemma nonsingularLift_zero [Nontrivial R] : W.NonsingularLift ⟦![1, 1, 0]⟧ :=
-  W.nonsingular_zero
-
-lemma nonsingularLift_zero' [NoZeroDivisors R] {Y : R} (hy : Y ≠ 0) :
-    W.NonsingularLift ⟦![Y ^ 2, Y ^ 3, 0]⟧ :=
-  W.nonsingular_zero' hy
```
[the remark this answers is not in the document]
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:274]: It's not fully removed - it's combined into the proof of `WeierstrassCurve.Jacobian.dblU_ne_zero_of_Y_eq`, and I don't see the use of this anymore. I don't mind adding it back somewhere, but do reuse it in the proof.
